File: crates/apollia-desktop/ui/src/lib/chat/confidenceParser.ts
/**
 * Confidence + citation markup parser (always-on).
 *
 * Mirrors `apollia-runtime::analyzers::confidence_parser`. Pure, zero deps.
 * Handles: `[conf:high|medium|low]…[/conf]` spans, nested, with inline
 * `[cite:<id>]` (or `[cite:a,b]`) pointers. Malformed markers are left
 * verbatim so the renderer simply falls back to classic prose.
 */

export type ConfidenceLevel = "high" | "medium" | "low";

export type SourceType = "web" | "document" | "tool" | "memory" | "other";

export interface Citation {
  id: string;
  title: string;
  url?: string;
  excerpt?: string;
  source_type?: SourceType;
}

export interface AssertionRange {
  start: number;
  end: number;
}

export interface Assertion {
  text_range: AssertionRange;
  confidence: ConfidenceLevel;
  citation_ids: string[];
}

export interface ParsedMessage {
  text: string;
  assertions: Assertion[];
}

interface OpenSpan {
  level: ConfidenceLevel;
  start: number;
  citationIds: string[];
}

type Token =
  | { kind: "open"; level: ConfidenceLevel }
  | { kind: "close" }
  | { kind: "cite"; ids: string[] };

const LEVELS: Record<string, ConfidenceLevel> = {
  high: "high",
  medium: "medium",
  med: "medium",
  low: "low",
};

function readToken(input: string, from: number): { token: Token; consumed: number } | null {
  const end = input.indexOf("]", from + 1);
  if (end === -1) return null;
  const inner = input.slice(from + 1, end);
  const consumed = end - from + 1;

  if (inner.startsWith("conf:")) {
    const raw = inner.slice(5).trim().toLowerCase();
    const level = LEVELS[raw];
    if (!level) return null;
    return { token: { kind: "open", level }, consumed };
  }
  if (inner === "/conf") {
    return { token: { kind: "close" }, consumed };
  }
  if (inner.startsWith("cite:")) {
    const ids = inner
      .slice(5)
      .split(",")
      .map((s) => s.trim())
      .filter((s) => s.length > 0);
    if (ids.length === 0) return null;
    return { token: { kind: "cite", ids }, consumed };
  }
  return null;
}

/**
 * Parse confidence/citation markers out of agent-produced text.
 *
 * Returns the cleaned text (markers stripped) plus the list of assertions
 * referencing character ranges in the cleaned text.
 */
interface ParserState {
  out: string;
  assertions: Assertion[];
  stack: OpenSpan[];
}

function applyOpen(state: ParserState, level: ConfidenceLevel): void {
  state.stack.push({ level, start: state.out.length, citationIds: [] });
}

/** Pop one span. Returns `true` when a span was actually closed. */
function applyClose(state: ParserState): boolean {
  const open = state.stack.pop();
  if (!open) return false;
  const endIdx = state.out.length;
  if (endIdx > open.start) {
    state.assertions.push({
      text_range: { start: open.start, end: endIdx },
      confidence: open.level,
      citation_ids: open.citationIds,
    });
  }
  return true;
}

/** Push citation ids onto the innermost open span. Returns `true` on success. */
function applyCite(state: ParserState, ids: string[]): boolean {
  const top = state.stack.at(-1);
  if (!top) return false;
  for (const id of ids) {
    if (!top.citationIds.includes(id)) {
      top.citationIds.push(id);
    }
  }
  return true;
}

/**
 * Apply a parsed token to the parser state. Returns `true` when the token
 * was consumed; `false` means the bracketed text must fall through as a
 * literal character (orphan close / orphan cite).
 */
function applyToken(state: ParserState, token: Token): boolean {
  if (token.kind === "open") {
    applyOpen(state, token.level);
    return true;
  }
  if (token.kind === "close") {
    return applyClose(state);
  }
  return applyCite(state, token.ids);
}

export function parseConfidence(raw: string): ParsedMessage {
  const state: ParserState = { out: "", assertions: [], stack: [] };
  let i = 0;

  while (i < raw.length) {
    const ch = raw[i];
    if (ch === "[") {
      const parsed = readToken(raw, i);
      if (parsed && applyToken(state, parsed.token)) {
        i += parsed.consumed;
        continue;
      }
    }
    state.out += ch;
    i += 1;
  }

  return { text: state.out, assertions: state.assertions };
}

/**
 * Split the cleaned text into a sequence of segments, each either
 * unannotated prose or a span belonging to exactly one assertion. Nested
 * assertions project onto the innermost span (renderer uses the top of the
 * nesting stack to pick the badge colour).
 */
export interface Segment {
  text: string;
  assertion: Assertion | null;
}

export function segmentize(parsed: ParsedMessage): Segment[] {
  if (parsed.assertions.length === 0) {
    return [{ text: parsed.text, assertion: null }];
  }
  // Collect boundary points and classify each character by the innermost
  // enclosing assertion. Innermost = the one with smallest span covering i.
  const n = parsed.text.length;
  const owner: (Assertion | null)[] = new Array(n).fill(null);
  // Sort assertions outer-first (larger span first); inner overwrites.
  const sorted = [...parsed.assertions].sort(
    (a, b) =>
      b.text_range.end - b.text_range.start - (a.text_range.end - a.text_range.start),
  );
  for (const a of sorted) {
    for (let k = a.text_range.start; k < a.text_range.end; k++) {
      owner[k] = a;
    }
  }
  const segs: Segment[] = [];
  let cursor = 0;
  while (cursor < n) {
    const current = owner[cursor];
    let j = cursor + 1;
    while (j < n && owner[j] === current) j++;
    segs.push({ text: parsed.text.slice(cursor, j), assertion: current });
    cursor = j;
  }
  return segs;
}

/** Resolve citation ids against a citation list, dropping unknown ids. */
export function resolveCitations(
  ids: string[],
  citations: Citation[],
): Citation[] {
  const byId = new Map(citations.map((c) => [c.id, c] as const));
  const out: Citation[] = [];
  for (const id of ids) {
    const c = byId.get(id);
    if (c) out.push(c);
  }
  return out;
}

File: crates/apollia-runtime/src/analyzers/confidence_parser.rs
//! Confidence + citation markup parser (US-SP42-046 — Pattern P10, always-on).
//!
//! Parses markers inline in agent output and returns a structured
//! [`ParsedMessage`] with assertions, citations and the cleaned text. No LLM
//! is involved — coût : 0.
//!
//! # Supported markers
//!
//! - `[conf:high]...[/conf]`, `[conf:medium]...[/conf]`, `[conf:low]...[/conf]`
//!   wrap a text span and tag it with a confidence level. Nested `conf`
//!   spans use the *innermost* level.
//! - `[cite:<id>]` is an inline pointer to a [`Citation`] in the message's
//!   citation list. Multiple citations inside the same `conf` span are
//!   attached to the enclosing assertion. `[cite:a,b]` shorthand is
//!   accepted and split.
//! - Citations themselves are declared out-of-band by the agent (in
//!   `AIPResult.metadata["citations"]`) or via the SDK helper.
//!
//! Malformed markers (unknown level, missing closing tag, orphan cite outside
//! a `conf` span) are left untouched in the cleaned text — the renderer
//! simply falls back to classic prose.

use serde::{Deserialize, Serialize};
use std::ops::Range;

/// Confidence level attached to an [`Assertion`].
///
/// Serialised in lower-case kebab form (`"high"`, `"medium"`, `"low"`).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ConfidenceLevel {
    High,
    Medium,
    Low,
}

impl ConfidenceLevel {
    fn parse(raw: &str) -> Option<Self> {
        match raw.trim().to_ascii_lowercase().as_str() {
            "high" => Some(Self::High),
            "medium" | "med" => Some(Self::Medium),
            "low" => Some(Self::Low),
            _ => None,
        }
    }
}

/// Broad classification of a citation's origin.
///
/// Drives the rendering of the footnote badge (icon, colour). Unknown
/// values serialise to `"other"`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "lowercase")]
pub enum SourceType {
    Web,
    Document,
    Tool,
    Memory,
    #[default]
    Other,
}

/// A bibliographic entry referenced by one or more [`Assertion`]s.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Citation {
    /// Stable identifier used in `[cite:<id>]` markers.
    pub id: String,
    /// Human-readable title shown in the footnote.
    pub title: String,
    /// Optional URL — rendered as a clickable link when present.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub url: Option<String>,
    /// Optional excerpt surfaced in the tooltip / side-panel.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub excerpt: Option<String>,
    /// Source classification.
    #[serde(default)]
    pub source_type: SourceType,
}

/// A span of the cleaned text tagged with a confidence level and citations.
///
/// `text_range` indexes into the cleaned (marker-stripped) text returned by
/// [`parse`]. Ranges are half-open byte ranges, following Rust conventions.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Assertion {
    pub text_range: AssertionRange,
    pub confidence: ConfidenceLevel,
    #[serde(default)]
    pub citation_ids: Vec<String>,
}

/// Serializable counterpart of `std::ops::Range<usize>` (serde-friendly).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct AssertionRange {
    pub start: usize,
    pub end: usize,
}

impl From<Range<usize>> for AssertionRange {
    fn from(r: Range<usize>) -> Self {
        Self {
            start: r.start,
            end: r.end,
        }
    }
}

impl From<AssertionRange> for Range<usize> {
    fn from(r: AssertionRange) -> Self {
        r.start..r.end
    }
}

/// Output of [`parse`].
///
/// `text` is the cleaned message (markers removed). `assertions` reference
/// ranges in `text`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ParsedMessage {
    pub text: String,
    #[serde(default)]
    pub assertions: Vec<Assertion>,
}

/// Parse confidence/citation markers out of *raw* agent text.
///
/// Unknown, malformed, or unbalanced markers are preserved verbatim so the
/// renderer can still show the underlying prose. Returns a
/// [`ParsedMessage`] with the cleaned text and recognised assertions.
pub fn parse(raw: &str) -> ParsedMessage {
    let mut out = String::with_capacity(raw.len());
    let mut assertions: Vec<Assertion> = Vec::new();
    let mut stack: Vec<OpenSpan> = Vec::new();
    let bytes = raw.as_bytes();
    let mut i = 0;

    while i < bytes.len() {
        if bytes[i] == b'[' {
            if let Some((token, consumed)) = read_token(&raw[i..]) {
                match token {
                    Token::OpenConf(level) => {
                        stack.push(OpenSpan {
                            level,
                            start: out.len(),
                            citation_ids: Vec::new(),
                        });
                        i += consumed;
                        continue;
                    }
                    Token::CloseConf => {
                        if let Some(open) = stack.pop() {
                            let end = out.len();
                            if end > open.start {
                                assertions.push(Assertion {
                                    text_range: (open.start..end).into(),
                                    confidence: open.level,
                                    citation_ids: open.citation_ids,
                                });
                            }
                            i += consumed;
                            continue;
                        }
                        // Orphan [/conf] → fall through as literal.
                    }
                    Token::Cite(ids) => {
                        if let Some(top) = stack.last_mut() {
                            for id in ids {
                                if !top.citation_ids.contains(&id) {
                                    top.citation_ids.push(id);
                                }
                            }
                            i += consumed;
                            continue;
                        }
                        // Orphan [cite:...] → keep as literal.
                    }
                }
            }
        }
        // Default: copy one UTF-8 character verbatim.
        let ch_len = utf8_char_len(bytes[i]);
        out.push_str(&raw[i..i + ch_len]);
        i += ch_len;
    }

    // Any still-open spans are malformed — we keep whatever text was emitted
    // but do NOT record an assertion (the closing marker was missing).

    ParsedMessage {
        text: out,
        assertions,
    }
}

// ---------------------------------------------------------------------------
// Internals
// ---------------------------------------------------------------------------

struct OpenSpan {
    level: ConfidenceLevel,
    start: usize,
    citation_ids: Vec<String>,
}

enum Token {
    OpenConf(ConfidenceLevel),
    CloseConf,
    Cite(Vec<String>),
}

/// Attempt to read a marker starting at the `[` character.
///
/// Returns the parsed token and the number of bytes consumed, or `None`
/// when the bracketed content is not a recognised marker.
fn read_token(input: &str) -> Option<(Token, usize)> {
    debug_assert!(input.starts_with('['));
    let end = input.find(']')?;
    let inner = &input[1..end];
    let consumed = end + 1;

    if let Some(rest) = inner.strip_prefix("conf:") {
        let level = ConfidenceLevel::parse(rest)?;
        return Some((Token::OpenConf(level), consumed));
    }
    if inner == "/conf" {
        return Some((Token::CloseConf, consumed));
    }
    if let Some(rest) = inner.strip_prefix("cite:") {
        let ids: Vec<String> = rest
            .split(',')
            .map(|s| s.trim())
            .filter(|s| !s.is_empty())
            .map(|s| s.to_string())
            .collect();
        if ids.is_empty() {
            return None;
        }
        return Some((Token::Cite(ids), consumed));
    }
    None
}

/// Byte-length of the UTF-8 character beginning at `lead`.
fn utf8_char_len(lead: u8) -> usize {
    if lead < 0x80 {
        1
    } else if lead < 0xC0 {
        1 // continuation byte, treat defensively
    } else if lead < 0xE0 {
        2
    } else if lead < 0xF0 {
        3
    } else {
        4
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn range(a: &Assertion) -> Range<usize> {
        a.text_range.into()
    }

    /// GIVEN text with a single high-confidence span
    /// WHEN parse
    /// THEN one assertion is produced and markers are stripped.
    #[test]
    fn parses_single_high_confidence_span() {
        let p = parse("Hello [conf:high]world[/conf] !");
        assert_eq!(p.text, "Hello world !");
        assert_eq!(p.assertions.len(), 1);
        assert_eq!(p.assertions[0].confidence, ConfidenceLevel::High);
        assert_eq!(&p.text[range(&p.assertions[0])], "world");
    }

    /// GIVEN nested conf spans
    /// WHEN parse
    /// THEN both assertions are emitted in post-order with correct ranges.
    #[test]
    fn handles_nested_conf_spans() {
        let p = parse("[conf:low]A [conf:high]B[/conf] C[/conf]");
        assert_eq!(p.text, "A B C");
        assert_eq!(p.assertions.len(), 2);
        // Inner closes first → pushed first.
        assert_eq!(p.assertions[0].confidence, ConfidenceLevel::High);
        assert_eq!(&p.text[range(&p.assertions[0])], "B");
        assert_eq!(p.assertions[1].confidence, ConfidenceLevel::Low);
        assert_eq!(&p.text[range(&p.assertions[1])], "A B C");
    }

    /// GIVEN a conf span with an inline cite
    /// WHEN parse
    /// THEN the cite id is attached to the assertion and removed from text.
    #[test]
    fn attaches_citations_to_enclosing_span() {
        let p = parse("Earth is round [conf:high]per NASA[cite:nasa-2024][/conf].");
        assert_eq!(p.text, "Earth is round per NASA.");
        let a = &p.assertions[0];
        assert_eq!(a.citation_ids, vec!["nasa-2024"]);
    }

    /// GIVEN multi-id cite shorthand
    /// WHEN parse
    /// THEN ids are split and deduplicated.
    #[test]
    fn parses_comma_separated_citation_ids() {
        let p = parse("[conf:medium]claim[cite:a, b,a][/conf]");
        assert_eq!(p.text, "claim");
        assert_eq!(p.assertions[0].citation_ids, vec!["a", "b"]);
    }

    /// GIVEN text with no markers
    /// WHEN parse
    /// THEN the output is identical and no assertions are produced.
    #[test]
    fn leaves_unmarked_text_unchanged() {
        let p = parse("Just a plain sentence.");
        assert_eq!(p.text, "Just a plain sentence.");
        assert!(p.assertions.is_empty());
    }

    /// GIVEN an unknown confidence level
    /// WHEN parse
    /// THEN the marker is preserved verbatim.
    #[test]
    fn unknown_level_is_left_verbatim() {
        let p = parse("[conf:ultra]foo[/conf]");
        assert_eq!(p.text, "[conf:ultra]foo[/conf]");
        assert!(p.assertions.is_empty());
    }

    /// GIVEN an orphan `[/conf]`
    /// WHEN parse
    /// THEN it is left verbatim and no assertion is emitted.
    #[test]
    fn orphan_close_is_left_verbatim() {
        let p = parse("foo[/conf] bar");
        assert_eq!(p.text, "foo[/conf] bar");
        assert!(p.assertions.is_empty());
    }

    /// GIVEN a conf span that is never closed
    /// WHEN parse
    /// THEN no assertion is emitted, inner text survives.
    #[test]
    fn unclosed_span_drops_assertion() {
        let p = parse("[conf:high]dangling");
        assert_eq!(p.text, "dangling");
        assert!(p.assertions.is_empty());
    }

    /// GIVEN an orphan `[cite:x]` outside any conf span
    /// WHEN parse
    /// THEN it is preserved as literal text.
    #[test]
    fn orphan_cite_is_literal() {
        let p = parse("see [cite:x]");
        assert_eq!(p.text, "see [cite:x]");
        assert!(p.assertions.is_empty());
    }

    /// GIVEN malformed brackets (missing `]`)
    /// WHEN parse
    /// THEN the `[` is copied verbatim and parsing continues.
    #[test]
    fn malformed_bracket_is_passthrough() {
        let p = parse("an [unbounded marker");
        assert_eq!(p.text, "an [unbounded marker");
    }

    /// GIVEN a conf span around multi-byte UTF-8 text
    /// WHEN parse
    /// THEN the range slices back into valid UTF-8.
    #[test]
    fn preserves_utf8_ranges() {
        let p = parse("[conf:high]éléphant 🐘[/conf]");
        let a = &p.assertions[0];
        let slice = &p.text[range(a)];
        assert_eq!(slice, "éléphant 🐘");
    }

    /// GIVEN `med` alias
    /// WHEN parse
    /// THEN it maps to Medium.
    #[test]
    fn accepts_med_alias() {
        let p = parse("[conf:med]x[/conf]");
        assert_eq!(p.assertions[0].confidence, ConfidenceLevel::Medium);
    }

    /// GIVEN multiple cites inside a single conf span
    /// WHEN parse
    /// THEN all ids accumulate on the assertion.
    #[test]
    fn multiple_cites_accumulate() {
        let p = parse("[conf:high]A[cite:a] B[cite:b][/conf]");
        assert_eq!(p.text, "A B");
        assert_eq!(p.assertions[0].citation_ids, vec!["a", "b"]);
    }
}
